lib/extractor/oci-distribution-metadata.ts

```diff
@@ -0,0 +1,122 @@
+import isValidHost from "is-valid-host";
+
+type OCIDistributionMetadata = {
+  // Must be a valid host, including port if one was used to pull the image.
+  // Max size: 255 bytes.
+  registryHost: string;
+  // Must be a valid OCI repository namespace.
+  // Max size: 2048 bytes.
+  repository: string;
+  // Must match sha256:[a-f0-9]{64} (https://github.com/opencontainers/image-spec/blob/d60099175f88c47cd379c4738d158884749ed235/descriptor.md?plain=1#L143).
+  // Max size: 64 bytes.
+  manifestDigest: string;
+  // (Optional) Must match sha256:[a-f0-9]{64} (https://github.com/opencontainers/image-spec/blob/d60099175f88c47cd379c4738d158884749ed235/descriptor.md?plain=1#L143).
+  //Max size: 64 bytes.
+  indexDigest?: string;
+  // (Optional) Must match [a-zA-Z0-9_][a-zA-Z0-9._-]{0,127} (https://github.com/opencontainers/distribution-spec/blob/3940529fe6c0a068290b27fb3cd797cf0528bed6/spec.md?plain=1#L160).
+  // Max size: 127 bytes.
+  imageTag?: string;
+  // Must be a valid OCI Image architecure/os string (https://github.com/opencontainers/image-spec/blob/93f6e65855a1e046b42afbad0ad129a3d44dc971/config.md?plain=1#L103).
+  // Max size: 64 bytes.
+  platform: string;
+};
+
+type OCIDistributionMetadataConstructorInput = {
+  imageName: string,
+  manifestDigest: string,
+  indexDigest?: string,
+  platform: string,
+};
+
+export function constructOCIDisributionMetadata({ imageName, manifestDigest, indexDigest, platform }: OCIDistributionMetadataConstructorInput): OCIDistributionMetadata | undefined {  
+  let registryHost: string, repositoryAndReference: string;
+  let hostOmitted = false;
+  const imageNameSplit = imageName.split('/', 2);
+  if (imageNameSplit.length > 1) {
+    [registryHost, repositoryAndReference] = imageNameSplit;
+  } else {
+    registryHost= 'docker.io';
+    hostOmitted = true;
+    [repositoryAndReference] = imageNameSplit;
+  }
+  
+  let repository: string, imageTag: string | undefined;
+  if (repositoryAndReference.includes('@')) {
+    [repository] = repositoryAndReference.split('@', 2);
+  } else if (repositoryAndReference.includes(':')) {
+    [repository, imageTag] = repositoryAndReference.split(':', 2);
+  } else {
+    imageTag = 'latest';
+    [repository] = repositoryAndReference;
+  }  
+
+  if (hostOmitted && !repository.includes('/')) {
+    repository = 'library/' + repository;
```

Comment:
what does `library` mean here?

Reply:
When you docker pull short names like `ubuntu`, you're really asking for `docker.io/library/ubuntu`.